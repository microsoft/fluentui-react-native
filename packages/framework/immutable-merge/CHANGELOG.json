{
  "name": "@uifabricshared/immutable-merge",
  "entries": [
    {
      "date": "Wed, 29 Jul 2020 22:38:20 GMT",
      "tag": "@uifabricshared/immutable-merge_v0.5.67",
      "version": "0.5.67",
      "comments": {
        "none": [
          {
            "comment": "switch to using Lage for build sequencing",
            "author": "jasonmo@microsoft.com",
            "commit": "79677c8f3260dbf55e053db5983fb161c6fb44b5",
            "package": "@uifabricshared/immutable-merge"
          }
        ]
      }
    },
    {
      "date": "Thu, 23 Jul 2020 21:28:55 GMT",
      "tag": "@uifabricshared/immutable-merge_v0.5.67",
      "version": "0.5.67",
      "comments": {
        "none": [
          {
            "comment": "Revert \"Change files\"",
            "author": "email not defined",
            "commit": "bac059cfd36f5a43185f3624e60bdda2d318c1b9",
            "package": "@uifabricshared/immutable-merge"
          }
        ],
        "patch": [
          {
            "comment": "fix comment",
            "author": "taamireh@microsoft.com",
            "commit": "c182f94c401e2d684e52b398992eecf3208979ac",
            "package": "@uifabricshared/immutable-merge"
          }
        ]
      }
    },
    {
      "date": "Thu, 02 Jul 2020 21:06:06 GMT",
      "tag": "@uifabricshared/immutable-merge_v0.5.66",
      "version": "0.5.66",
      "comments": {
        "none": [
          {
            "comment": "Revert \"Change files\"",
            "author": "email not defined",
            "commit": "bac059cfd36f5a43185f3624e60bdda2d318c1b9",
            "package": "@uifabricshared/immutable-merge"
          }
        ]
      }
    },
    {
      "date": "Wed, 17 Jun 2020 05:00:56 GMT",
      "tag": "@uifabricshared/immutable-merge_v0.5.57",
      "version": "0.5.57",
      "comments": {
        "patch": [
          {
            "comment": "add memo-cache package and consume it in the core framework",
            "author": "jasonmo@microsoft.com",
            "commit": "27b5fc6790369ab629dcf7f93d6091d4913915fa",
            "package": "@uifabricshared/immutable-merge"
          }
        ]
      }
    },
    {
      "date": "Thu, 11 Jun 2020 23:53:15 GMT",
      "tag": "@uifabricshared/immutable-merge_v0.5.44",
      "version": "0.5.44",
      "comments": {
        "none": [
          {
            "comment": "Merge branch 'master' into button-tooltip",
            "author": "lehon@microsoft.com",
            "commit": "70f7d6d5972b384f2e8d4dd2626adda762c59206",
            "package": "@uifabricshared/immutable-merge"
          }
        ]
      }
    },
    {
      "date": "Tue, 09 Jun 2020 17:39:01 GMT",
      "tag": "@uifabricshared/immutable-merge_v0.5.39",
      "version": "0.5.39",
      "comments": {
        "patch": [
          {
            "comment": "switch dependencies to use greater than semver until we reach 1.0.0",
            "author": "jasonmo@microsoft.com",
            "commit": "8461668e0387928c0831846b219188a6d4408295",
            "package": "@uifabricshared/immutable-merge"
          }
        ]
      }
    },
    {
      "date": "Sat, 06 Jun 2020 03:03:17 GMT",
      "tag": "@uifabricshared/immutable-merge_v0.5.37",
      "version": "0.5.37",
      "comments": {
        "none": [
          {
            "comment": "support keyboardFocus for Link and add focus visuals",
            "author": "lehon@microsoft.com",
            "commit": "effb17ada2fd26c6ba8f7cbb4ab7d7ddf8d74c49",
            "package": "@uifabricshared/immutable-merge"
          }
        ]
      }
    },
    {
      "date": "Mon, 01 Jun 2020 22:06:07 GMT",
      "tag": "@uifabricshared/immutable-merge_v0.5.33",
      "version": "0.5.33",
      "comments": {
        "none": [
          {
            "comment": "Support enter key for calling onClick on Button",
            "author": "lehon@microsoft.com",
            "commit": "399cead569e120926521d01c13a49088fe96036b",
            "package": "@uifabricshared/immutable-merge"
          }
        ]
      }
    },
    {
      "date": "Thu, 21 May 2020 20:00:42 GMT",
      "tag": "@uifabricshared/immutable-merge_v0.5.28",
      "version": "0.5.28",
      "comments": {
        "none": [
          {
            "comment": "none",
            "author": "lehon@microsoft.com",
            "commit": "5ba3ac570ae17a0d3138302ad12807362dea06c1",
            "package": "@uifabricshared/immutable-merge"
          }
        ]
      }
    },
    {
      "date": "Wed, 20 May 2020 03:37:41 GMT",
      "tag": "@uifabricshared/immutable-merge_v0.5.25",
      "version": "0.5.25",
      "comments": {
        "none": [
          {
            "comment": "ðŸ“¦ applying package updates ***NO_CI***",
            "author": "lehon@microsoft.com",
            "commit": "4867e1dde6f8c5cdb3fe894e60f5bcbe8f3c2817",
            "package": "@uifabricshared/immutable-merge"
          }
        ]
      }
    },
    {
      "date": "Tue, 19 May 2020 20:06:57 GMT",
      "tag": "@uifabricshared/immutable-merge_v0.5.22",
      "version": "0.5.22",
      "comments": {
        "none": [
          {
            "comment": "ðŸ”€ Missed file in merge",
            "author": "email not defined",
            "commit": "16687f10ee6e6115cb0756e936eecacb4e198b85",
            "package": "@uifabricshared/immutable-merge"
          }
        ]
      }
    },
    {
      "date": "Fri, 08 May 2020 18:49:53 GMT",
      "tag": "@uifabricshared/immutable-merge_v0.5.13",
      "version": "0.5.13",
      "comments": {
        "none": [
          {
            "comment": "ðŸ“¦ applying package updates ***NO_CI***",
            "author": "lehon@microsoft.com",
            "commit": "ff5a5b0edb37bb35c9b2a4e817089a5a56551604",
            "package": "@uifabricshared/immutable-merge"
          }
        ]
      }
    },
    {
      "date": "Tue, 21 Apr 2020 05:58:24 GMT",
      "tag": "@uifabricshared/immutable-merge_v0.5.2",
      "version": "0.5.2",
      "comments": {
        "patch": [
          {
            "comment": "pick up patched beachball version",
            "author": "jasonmo@microsoft.com",
            "commit": "68d5e18fc919616cbb34fb090936e4281f52609f",
            "package": "@uifabricshared/immutable-merge"
          }
        ]
      }
    },
    {
      "date": "Tue, 21 Apr 2020 00:41:10 GMT",
      "tag": "@uifabricshared/immutable-merge_v0.5.1",
      "version": "0.5.1",
      "comments": {
        "patch": [
          {
            "comment": "update beachball hook to use new prepublish strategy",
            "author": "jasonmo@microsoft.com",
            "commit": "7bc030df722a6fcb6f693658b9d80f4617a26713",
            "package": "@uifabricshared/immutable-merge"
          }
        ]
      }
    },
    {
      "date": "Fri, 17 Apr 2020 22:36:03 GMT",
      "tag": "@uifabricshared/immutable-merge_v0.5.0",
      "version": "0.5.0",
      "comments": {
        "minor": [
          {
            "comment": "yarn.lock",
            "author": "warleu@microsoft.com",
            "commit": "205e7b7e88151a17f2560ecdaec1c6e13c5825b7",
            "package": "@uifabricshared/immutable-merge"
          }
        ]
      }
    },
    {
      "date": "Thu, 16 Apr 2020 23:57:38 GMT",
      "tag": "@uifabricshared/immutable-merge_v0.4.1",
      "version": "0.4.1",
      "comments": {
        "patch": [
          {
            "comment": "publish with correct main/module references",
            "author": "jasonmo@microsoft.com",
            "commit": "6709e9eed6e08029114e8d893d27d91b99f73d0a",
            "package": "@uifabricshared/immutable-merge"
          }
        ]
      }
    },
    {
      "date": "Thu, 09 Apr 2020 21:14:58 GMT",
      "tag": "@uifabricshared/immutable-merge_v0.4.0",
      "version": "0.4.0",
      "comments": {
        "minor": [
          {
            "comment": "rework merge with improved API",
            "author": "jasonmo@microsoft.com",
            "commit": "232a97794dc044cec9671e5cca4cf204d881f614",
            "package": "@uifabricshared/immutable-merge"
          }
        ]
      }
    },
    {
      "date": "Thu, 09 Apr 2020 18:39:15 GMT",
      "tag": "@uifabricshared/immutable-merge_v0.3.5",
      "version": "0.3.5",
      "comments": {
        "patch": [
          {
            "comment": "merge conflicts",
            "author": "ppatboyd@outlook.com",
            "commit": "232a97794dc044cec9671e5cca4cf204d881f614",
            "package": "@uifabricshared/immutable-merge"
          }
        ]
      }
    },
    {
      "date": "Fri, 03 Apr 2020 20:40:51 GMT",
      "tag": "@uifabricshared/immutable-merge_v0.3.3",
      "version": "0.3.3",
      "comments": {
        "patch": [
          {
            "comment": "switch the bin name from just-script to fluent-scripts to disambiguate names",
            "author": "jasonmo@microsoft.com",
            "commit": "232a97794dc044cec9671e5cca4cf204d881f614",
            "package": "@uifabricshared/immutable-merge"
          }
        ]
      }
    },
    {
      "date": "Tue, 31 Mar 2020 18:01:09 GMT",
      "tag": "@uifabricshared/immutable-merge_v0.3.2",
      "version": "0.3.2",
      "comments": {
        "patch": [
          {
            "comment": "Merge branch 'sammy/checkboxV1' of https://github.com/samuelfreiberg/fluentui-react-native into sammy/checkboxV1",
            "author": "safreibe@microsoft.com",
            "commit": "d38c7c235ebfe8278a31a4318d14ba76a98ef2a6"
          }
        ]
      }
    },
    {
      "date": "Mon, 30 Mar 2020 16:27:00 GMT",
      "tag": "@uifabricshared/immutable-merge_v0.3.1",
      "version": "0.3.1",
      "comments": {
        "patch": [
          {
            "comment": "Update main attribute to specify TypeScript index file.",
            "author": "kinhln@microsoft.com",
            "commit": "ab291fb90d584c81310aa6559d05a360e86f97de"
          }
        ]
      }
    },
    {
      "date": "Fri, 27 Mar 2020 04:10:56 GMT",
      "tag": "@uifabricshared/immutable-merge_v0.3.0",
      "version": "0.3.0",
      "comments": {
        "none": [
          {
            "comment": "shared package configs",
            "author": "ppatboyd@outlook.com",
            "commit": "1471793de9a859b9908ef50069aeb673d06e5b42"
          }
        ]
      }
    },
    {
      "date": "Sat, 07 Mar 2020 00:42:50 GMT",
      "tag": "@uifabricshared/immutable-merge_v0.3.0",
      "version": "0.3.0",
      "comments": {
        "minor": [
          {
            "comment": "more config fixes",
            "author": "jasonmo@microsoft.com",
            "commit": "bdb2a853738f6b4819bc388c351599c982ac2790"
          }
        ]
      }
    },
    {
      "date": "Thu, 20 Feb 2020 01:36:35 GMT",
      "tag": "@uifabricshared/immutable-merge_v0.2.3",
      "version": "0.2.3",
      "comments": {
        "patch": [
          {
            "comment": "restructure directories in project",
            "author": "jasonmo@microsoft.com",
            "commit": "6a2d86027f68f50423fed6d893fcc8071e67e89c"
          }
        ]
      }
    },
    {
      "date": "Fri, 07 Feb 2020 00:29:25 GMT",
      "tag": "@uifabricshared/immutable-merge_v0.2.2",
      "version": "0.2.2",
      "comments": {
        "patch": [
          {
            "comment": "clean up just tasks and remove unnecessary webpack configs",
            "author": "jasonmo@microsoft.com",
            "commit": "325ea8c30a8c8fbe37f84cd5b5080fc7e6c2864b"
          }
        ]
      }
    },
    {
      "date": "Thu, 06 Feb 2020 01:00:13 GMT",
      "tag": "@uifabricshared/immutable-merge_v0.2.1",
      "version": "0.2.1",
      "comments": {
        "patch": [
          {
            "comment": "clean up just tasks and remove unnecessary webpack configs",
            "author": "jasonmo@microsoft.com",
            "commit": "325ea8c30a8c8fbe37f84cd5b5080fc7e6c2864b"
          }
        ]
      }
    },
    {
      "date": "Wed, 20 Nov 2019 19:37:06 GMT",
      "tag": "@uifabricshared/immutable-merge_v0.2.0",
      "version": "0.2.0",
      "comments": {
        "minor": [
          {
            "comment": "ðŸ“¦ fix package entrypoints to distinguish CJS/ES6",
            "author": "adrum@microsoft.com",
            "commit": "d51fcaea614de5606989d8463c7f66ccbc5322cf"
          }
        ]
      }
    },
    {
      "date": "Tue, 15 Oct 2019 02:11:41 GMT",
      "tag": "@uifabricshared/immutable-merge_v0.1.6",
      "version": "0.1.6",
      "comments": {
        "patch": [
          {
            "comment": "api-extractor setup",
            "author": "taamireh@microsoft.com",
            "commit": "eb1d0cf52629e6efe6241776cc97a25585e866a4"
          }
        ]
      }
    },
    {
      "date": "Wed, 25 Sep 2019 23:15:41 GMT",
      "tag": "@uifabricshared/immutable-merge_v0.1.5",
      "version": "0.1.5",
      "comments": {
        "none": [
          {
            "comment": "Disallow major changes",
            "author": "taamireh@microsoft.com",
            "commit": "30c4e1f286520bd98c5f4fcd2f83a2559d0a7e71"
          }
        ]
      }
    },
    {
      "date": "Thu, 19 Sep 2019 23:21:48 GMT",
      "tag": "@uifabricshared/immutable-merge_v0.1.5",
      "version": "0.1.5",
      "comments": {
        "none": [
          {
            "comment": "force publish",
            "author": "taamireh@microsoft.com",
            "commit": "c8c425c7532334f9a742980167c3c204732a5a1c"
          }
        ]
      }
    },
    {
      "date": "Thu, 19 Sep 2019 18:26:03 GMT",
      "tag": "@uifabricshared/immutable-merge_v0.1.5",
      "version": "0.1.5",
      "comments": {
        "patch": [
          {
            "comment": "change description",
            "author": "taamireh@microsoft.com",
            "commit": "3127d5c0b0fbb56bbaee00e0bb0cbe276cc48c4c"
          }
        ]
      }
    },
    {
      "date": "Thu, 19 Sep 2019 00:21:30 GMT",
      "tag": "@uifabricshared/immutable-merge_v0.1.4",
      "version": "0.1.4",
      "comments": {
        "patch": [
          {
            "comment": "change scope names",
            "author": "taamireh@microsoft.com",
            "commit": "a1e4beb10c154da04dd03f4588fa5422e51d5edc"
          }
        ]
      }
    },
    {
      "date": "Sat, 14 Sep 2019 01:56:36 GMT",
      "tag": "@uifabric/immutable-merge_v0.1.3",
      "version": "0.1.3",
      "comments": {
        "none": [
          {
            "comment": "update yarn.lock",
            "author": "taamireh@microsoft.com",
            "commit": "b28351081d8cb19c8f9bec9bc8f7f606eb2b251e"
          }
        ],
        "patch": [
          {
            "comment": "Begin turning on some linting rules",
            "author": "taamireh@microsoft.com",
            "commit": "530f021858edbb964892dffcb6f15f86db8499fd"
          }
        ]
      }
    },
    {
      "date": "Thu, 12 Sep 2019 20:46:12 GMT",
      "tag": "@uifabric/immutable-merge_v0.1.2",
      "version": "0.1.2",
      "comments": {
        "patch": [
          {
            "comment": "Add exploratory item to immutable merge",
            "author": "taamireh@microsoft.com",
            "commit": "eadfcb234f79cfc208431571572e06ccbf00efac"
          }
        ]
      }
    }
  ]
}
