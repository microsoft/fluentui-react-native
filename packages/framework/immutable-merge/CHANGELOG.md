# Change Log - @uifabricshared/immutable-merge

This log was last generated on Thu, 23 Jul 2020 21:28:55 GMT and should not be manually modified.

<!-- Start content -->

## 0.5.67

Thu, 23 Jul 2020 21:28:55 GMT

### Patches

- fix comment (taamireh@microsoft.com)

## 0.5.57

Wed, 17 Jun 2020 05:00:56 GMT

### Patches

- add memo-cache package and consume it in the core framework (jasonmo@microsoft.com)

## 0.5.39

Tue, 09 Jun 2020 17:39:01 GMT

### Patches

- switch dependencies to use greater than semver until we reach 1.0.0 (jasonmo@microsoft.com)

## 0.5.2

Tue, 21 Apr 2020 05:58:24 GMT

### Patches

- pick up patched beachball version (jasonmo@microsoft.com)

## 0.5.1

Tue, 21 Apr 2020 00:41:10 GMT

### Patches

- update beachball hook to use new prepublish strategy (jasonmo@microsoft.com)

## 0.5.0

Fri, 17 Apr 2020 22:36:03 GMT

### Minor changes

- yarn.lock (warleu@microsoft.com)

## 0.4.1

Thu, 16 Apr 2020 23:57:38 GMT

### Patches

- publish with correct main/module references (jasonmo@microsoft.com)

## 0.4.0
Thu, 09 Apr 2020 21:14:58 GMT

### Minor changes

- rework merge with improved API (jasonmo@microsoft.com)
## 0.3.5
Thu, 09 Apr 2020 18:39:15 GMT

### Patches

- merge conflicts (ppatboyd@outlook.com)
## 0.3.3
Fri, 03 Apr 2020 20:40:51 GMT

### Patches

- switch the bin name from just-script to fluent-scripts to disambiguate names (jasonmo@microsoft.com)
## 0.3.2
Tue, 31 Mar 2020 18:01:09 GMT

### Patches

- Merge branch 'sammy/checkboxV1' of https://github.com/samuelfreiberg/fluentui-react-native into sammy/checkboxV1 (safreibe@microsoft.com)
## 0.3.1
Mon, 30 Mar 2020 16:27:00 GMT

### Patches

- Update main attribute to specify TypeScript index file. (kinhln@microsoft.com)
## 0.3.0
Sat, 07 Mar 2020 00:42:50 GMT

### Minor changes

- more config fixes (jasonmo@microsoft.com)
## 0.2.3
Thu, 20 Feb 2020 01:36:35 GMT

### Patches

- restructure directories in project (jasonmo@microsoft.com)
## 0.2.2
Fri, 07 Feb 2020 00:29:25 GMT

### Patches

- clean up just tasks and remove unnecessary webpack configs (jasonmo@microsoft.com)
## 0.2.1
Thu, 06 Feb 2020 01:00:13 GMT

### Patches

- clean up just tasks and remove unnecessary webpack configs (jasonmo@microsoft.com)
## 0.2.0
Wed, 20 Nov 2019 19:37:06 GMT

### Minor changes

- ðŸ“¦ fix package entrypoints to distinguish CJS/ES6 (adrum@microsoft.com)
## 0.1.6
Tue, 15 Oct 2019 02:11:41 GMT

### Patches

- api-extractor setup (taamireh@microsoft.com)
## 0.1.5
Thu, 19 Sep 2019 18:26:03 GMT

### Patches

- change description (taamireh@microsoft.com)
## 0.1.4
Thu, 19 Sep 2019 00:21:30 GMT

### Patches

- change scope names (taamireh@microsoft.com)
## 0.1.3
Sat, 14 Sep 2019 01:56:36 GMT

### Patches

- Begin turning on some linting rules (taamireh@microsoft.com)
## 0.1.2
Thu, 12 Sep 2019 20:46:12 GMT

### Patches

- Add exploratory item to immutable merge (taamireh@microsoft.com)
