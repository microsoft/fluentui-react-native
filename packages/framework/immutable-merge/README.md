# `@fluentui-react-native/immutable-merge`

This package is now a pass-through package which references `@fluentui-react-native/framework-base` for backwards compatibility.

The current documentation can be found [here](../../framework-base/src/immutable-merge/README.md).
