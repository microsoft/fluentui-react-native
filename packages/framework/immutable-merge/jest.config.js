const { configureJest } = require('@fluentui-react-native/scripts');
module.exports = configureJest({
  // Jest is picking up the hoisted version of lru-cache, which is incompatible from the version
  // required by semver
  moduleNameMapper: {
    'lru-cache': require.resolve('lru-cache', {paths:[require.resolve('semver')]}),
  },
});