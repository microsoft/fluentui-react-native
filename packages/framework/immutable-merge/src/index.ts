export { immutableMerge, immutableMergeCore, processImmutable } from './Merge';
export type { BuiltinRecursionHandlers, CustomRecursionHandler, MergeOptions, RecursionHandler, RecursionOption } from './Merge';
