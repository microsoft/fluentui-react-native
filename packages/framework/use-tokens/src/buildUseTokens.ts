import { getMemoCache, GetMemoValue } from '@fluentui-react-native/memo-cache';
import { immutableMerge } from '@fluentui-react-native/immutable-merge';

/** A function to generate tokens based on a theme */
export type TokensFromTheme<TTokens, TTheme> = (theme: TTheme) => TTokens;

/**
 * Types of tokens, can be:
 * - string - will lookup the name in the theme
 * - Tokens - will merge the tokens in directly
 * - Function - will run against the theme once for each unique theme encountered
 */
export type TokenSettings<TTokens, TTheme> = string | TTokens | TokensFromTheme<TTokens, TTheme>;

/**
 * The main signature of a useTokens hook is to take the theme and produce a set of resolved tokens,
 * as well as a sub-cache, specific to this particular theme, that can be used for caching various styles
 * or values that are theme specific
 */
export type UseTokensCore<TTokens, TTheme> = (theme: TTheme) => [TTokens, GetMemoValue<TTokens>];

/**
 * The full signature also includes a customize function that returns an updated version of useTokens
 * that captures both the previous values, and layers in the new values specified
 */
export type UseTokens<TTokens, TTheme> = UseTokensCore<TTokens, TTheme> & {
  customize: (...tokens: TokenSettings<TTokens, TTheme>[]) => UseTokens<TTokens, TTheme>;
};

/**
 * Helper function that knows how to try to look up token information from the theme
 */
export type GetComponentInfo<TTokens, TTheme> = (theme: TTheme, name: string) => TTokens | TokensFromTheme<TTokens, TTheme>;

/**
 * Tokens are defined as either:
 *   TTokens     - an object
 *   string      - a name to look up in the theme
 *   function    - a function to run against the theme to produce tokens
 *
 * This function maps any of these types into a specific TTokens object.  A string is first lookup up in the theme, returning a function
 * or object. If the type is a function this will be invoked with the theme to generate the tokens object.
 *
 * @param tokenEntry - token entry to start with
 * @param theme - theme to use for queries
 * @param getComponentInfo - helper to use to lookup the component in the theme
 */
function mapToTokens<TTokens, TTheme>(
  tokenEntry: TTokens | string | TokensFromTheme<TTokens, TTheme>,
  theme: TTheme,
  getComponentInfo: GetComponentInfo<TTokens, TTheme> | undefined,
): object {
  if (typeof tokenEntry === 'string') {
    tokenEntry = (getComponentInfo && (getComponentInfo(theme, tokenEntry) as TTokens)) || ({} as TTokens);
  }
  if (typeof tokenEntry === 'function') {
    tokenEntry = (tokenEntry as TokensFromTheme<TTokens, TTheme>)(theme);
  }
  return tokenEntry as unknown as object;
}

/**
 * Construct a useStyling hook which returns styled slot props based on props and tokens defined in options and in the theme
 *
 * @param options - options which drive behavior for the generated styling hook
 * @param themeHelper - injected theme functionality
 */
export function buildUseTokens<TTokens, TTheme>(
  getComponentInfo: GetComponentInfo<TTokens, TTheme> | undefined,
  ...tokens: TokenSettings<TTokens, TTheme>[]
): UseTokens<TTokens, TTheme> {
  // create a cache instance for use in this particular call to buildUseTokens
  const cache = getMemoCache();

  // the core function simply merges layers together, looking up component definitions in the theme as well as executing any
  // theme functions. This turns the tokens into an array of token objects that then get merged together
  const useTokensCore = (theme: TTheme) => {
    // get the base styles all merged together, these will only depend on internal tokens and theme
    return cache(() => immutableMerge(...tokens.map((value) => mapToTokens(value, theme, getComponentInfo))), [theme]);
  };

  // attach a customize function to generate a new use
  useTokensCore.customize = (...newTokens: TokenSettings<TTokens, TTheme>[]) => {
    const mergedTokens = [...tokens, ...newTokens];
    return buildUseTokens<TTokens, TTheme>(getComponentInfo, ...mergedTokens);
  };

  return useTokensCore;
}
