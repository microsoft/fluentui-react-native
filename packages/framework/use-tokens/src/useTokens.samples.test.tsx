import { TextProps, Text, View } from 'react-native';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import { mount } from 'enzyme';
import { buildUseTokens } from './buildUseTokens';
import { mergeStyles } from '@fluentui-react-native/merge-props';
import { immutableMerge } from '@fluentui-react-native/immutable-merge';

/**
 * Sample super simple theming implementation, shared by all the samples. This is intended to be illustrative,
 * not to be production ready theming code
 */

/**
 * The theme just contains global values and options for looking up component overrides
 */
type Theme = {
  globals: {
    backgroundColor: string;
    color: string;
    borderColor: string;
    fontFamily: string;
    fontSize: number;
  };
  components: {
    [key: string]: any;
  };
};

/**
 * The default/base theme just contains base values
 */
const baseTheme: Theme = {
  globals: {
    backgroundColor: 'white',
    color: 'black',
    borderColor: 'blue',
    fontFamily: 'Arial',
    fontSize: 12,
  },
  components: {},
};

const current = { theme: baseTheme };

const useTheme = () => current.theme;

const setActiveTheme = (theme?: Partial<Theme>) => {
  current.theme = (theme && immutableMerge<Theme>(baseTheme, theme as Theme)) || baseTheme;
};

/**
 * this wrapper solves the (so-far) inexplicable type errors from the matchers in typescript
 */
function snapshotTestTree(tree: any) {
  (expect(toJson(tree)) as any).toMatchSnapshot();
}

/**
 * Helper function used to look up a component in the theme. Having this injected allows this module to not be dependent on the shape of
 * the theme used.
 */
const getComponentInfo = (theme: Theme, name: string) => theme.components[name];

describe('useTokens samples', () => {
  /**
   * Sample #1 - Themeable text element
   *
   * This adds some default opinions for how a text element should be styled but only allows for customization
   * via theming
   */

  // the tokens for customization are just the color and font props from the theme
  type Sample1Tokens = {
    color?: string;
    fontFamily?: string;
    fontSize?: number;
  };

  const useTokensSample1 = buildUseTokens<Sample1Tokens, Theme>(
    getComponentInfo,
    /** first the default values should come from the global theme section */
    (t: Theme) => ({
      color: t.globals.color,
      fontFamily: t.globals.fontFamily,
      fontSize: t.globals.fontSize,
    }),
    /** next we should look for a component reference to overlay */
    'SampleText',
  );

  const SampleText1: React.FunctionComponent<TextProps> = (props) => {
    // standard props splitting
    const { style, children, ...rest } = props;

    // typically this would start with a call to retrieve the theme from the context via whatever method is appropriate
    const theme = useTheme();

    // next the tokens are resolved from the theme, a cache specific to this theme is returned as well to allow for
    // style objects to not be rebuilt unnecessarily
    const [tokens, cache] = useTokensSample1(theme);

    // build up the text style, or the full props as appropriate
    const styleFromTokens = cache(
      /**
       * first build the style object
       * - this executes once for every unique set of keys.
       * - The cache is already unique for this theme
       */
      () => ({ ...tokens }),
      /**
       * now specify the keys
       * - because the only changing variable is the theme
       * - ...and all the style properties are either constant or come from the tokens
       * - ...no keys need to be specified
       * - this means that only one style object will be created per component + theme pair
       */
      [],
    );

    // merge the props from the tokens with anything passed in via style. This is internally cached via object identity
    // so the merged style object won't change identity unless one of the two inputs changes identity.
    const mergedStyle = mergeStyles(styleFromTokens, style);

    // now just render the element, forwarding the props, setting the merged style, then passing the children as appropriate
    return (
      <Text {...rest} style={mergedStyle}>
        {children}
      </Text>
    );
  };

  beforeEach(() => {
    setActiveTheme();
  });

  /** first render the component with no updates */
  it('Sample1Text rendering with no overrides', () => {
    const tree = mount(<SampleText1>Sample1a</SampleText1>);
    snapshotTestTree(tree);
  });

  /** now re-theme the component via the components in the theme */
  it('Sample1Text rendering with some custom settings in the theme', () => {
    setActiveTheme({
      components: {
        SampleText: {
          color: 'pink',
          fontSize: 24,
        },
      },
    });
    const tree = mount(<SampleText1>Sample1b</SampleText1>);
    snapshotTestTree(tree);
  });

  /**
   * Sample 2 - take the styled text component from the first example and allow color to be set in props as well. This is
   * effectively saying that all other styles come from the theme except that color can be overriden via props.
   */

  /** use the props from the first sample and just add a color setting */
  type Sample2Props = TextProps & { color?: string };

  // the Sample2Text component is built the same way as sample1, just using the new hook that has been created
  const SampleText2: React.FunctionComponent<Sample2Props> = (props) => {
    const { color, style, children, ...rest } = props;
    const theme = useTheme();

    // this starts the same as sample1, extract tokens from the theme and get a theme specific cache object
    const [tokens, cache] = useTokensSample1(theme);

    // now when building up the style this time, the resulting style object is based upon both the theme and the passed
    // in value of colors. Because the theme is already part of the cache definition, only color needs to be a key
    const styleFromTokens = cache(
      /** build the style, only patch the color if it has a value, otherwise the theme value would get stomped if color was undefined */
      () => ({ ...tokens, ...(color && { color }) }),
      /** use color as an additional key for the style */
      [color],
    );

    // now just render, this time merging styles inline to make it a bit shorter
    return (
      <Text {...rest} style={mergeStyles(styleFromTokens, style)}>
        {children}
      </Text>
    );
  };

  /** rendering the Sample2 component with the base theme */
  it('Sample2Text rendering with defaults and a color override', () => {
    const tree = mount(
      <View>
        <SampleText2>Sample2 with defaults</SampleText2>
        <SampleText2 color="green">Sample2 with color override via prop</SampleText2>
      </View>,
    );
    snapshotTestTree(tree);
  });

  /** now re-theme the component via the components in the theme */
  it('Sample2Text rendering with some custom settings in the theme', () => {
    setActiveTheme({
      components: {
        SampleText: {
          fontSize: 18,
          fontFamily: 'Helvetica',
        },
      },
    });
    const tree = mount(
      <View>
        <SampleText2>Sample2 with theme overrides set</SampleText2>
        <SampleText2 color="purple">Sample2 with theme and color prop override</SampleText2>
      </View>,
    );
    snapshotTestTree(tree);
  });
});
