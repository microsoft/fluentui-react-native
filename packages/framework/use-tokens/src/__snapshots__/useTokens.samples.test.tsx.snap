// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useTokens samples Sample1Text rendering with no overrides 1`] = `
<SampleText1>
  <Text
    style={
      Object {
        "color": "black",
        "fontFamily": "Arial",
        "fontSize": 12,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "black",
          "fontFamily": "Arial",
          "fontSize": 12,
        }
      }
    >
      Sample1a
    </Text>
  </Text>
</SampleText1>
`;

exports[`useTokens samples Sample1Text rendering with some custom settings in the theme 1`] = `
<SampleText1>
  <Text
    style={
      Object {
        "color": "pink",
        "fontFamily": "Arial",
        "fontSize": 24,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "pink",
          "fontFamily": "Arial",
          "fontSize": 24,
        }
      }
    >
      Sample1b
    </Text>
  </Text>
</SampleText1>
`;

exports[`useTokens samples Sample2Text rendering with defaults and a color override 1`] = `
<View>
  <View>
    <SampleText2>
      <Text
        style={
          Object {
            "color": "black",
            "fontFamily": "Arial",
            "fontSize": 12,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "black",
              "fontFamily": "Arial",
              "fontSize": 12,
            }
          }
        >
          Sample2 with defaults
        </Text>
      </Text>
    </SampleText2>
    <SampleText2
      color="green"
    >
      <Text
        style={
          Object {
            "color": "green",
            "fontFamily": "Arial",
            "fontSize": 12,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "green",
              "fontFamily": "Arial",
              "fontSize": 12,
            }
          }
        >
          Sample2 with color override via prop
        </Text>
      </Text>
    </SampleText2>
  </View>
</View>
`;

exports[`useTokens samples Sample2Text rendering with some custom settings in the theme 1`] = `
<View>
  <View>
    <SampleText2>
      <Text
        style={
          Object {
            "color": "black",
            "fontFamily": "Helvetica",
            "fontSize": 18,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "black",
              "fontFamily": "Helvetica",
              "fontSize": 18,
            }
          }
        >
          Sample2 with theme overrides set
        </Text>
      </Text>
    </SampleText2>
    <SampleText2
      color="purple"
    >
      <Text
        style={
          Object {
            "color": "purple",
            "fontFamily": "Helvetica",
            "fontSize": 18,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "purple",
              "fontFamily": "Helvetica",
              "fontSize": 18,
            }
          }
        >
          Sample2 with theme and color prop override
        </Text>
      </Text>
    </SampleText2>
  </View>
</View>
`;
